### PR TITLE
[config] Config support for including other config file

### PR DESCRIPTION
**Pull request by ridz1208**
projects can use the `<include>` tag to import secondary config xml files and append them to the main config array.

This can be used to separate the database and couchdb settings from the main config

### CODE REVIEW

**Comment by ridz1208** [comment]
Discussion:
- line 65/66 why is that if() there ?
- precedence problem with couchdb/database and database
- utility::toArray() bug https://github.com/aces/Loris/pull/2714

**Comment by ridz1208** [comment]
@kongtiaowang can we look at this one together when you have a minute ?

**Comment by ridz1208** [comment]
@ridz1208
will rename the couchDB database tag

**Comment by johnsaigle** [comment]
Added `Needs Work...` until database tag removed.

**Comment by kongtiaowang** [comment]
@ridz1208 I will figure it out.

**Comment by ridz1208** [comment]
@kongtiaowang any luck?

**Comment by kongtiaowang** [comment]
@ridz1208 
"$config->_settings['include']" is empty in travis_ci.
So I put  "if (!empty($config->_settings['include']))" before you run "$config->load($path)".
It works. Could u fix the format? It faild because PHPCS.

**Comment by kongtiaowang** [comment]
LGTM

**Comment by ridz1208** [comment]
thank you @kongtiaowang

**Comment by johnsaigle** [comment]
@driusan requested to review by @ridz1208

**Comment by ridz1208** [comment]
@xlecours would you be interested in testing/using this ??

**Comment by AnyhowStep** [comment]
The relative root directory, when `config.xml` is parsed, is `htdocs`. Probably not desirable. It should be the same directory as the "main" `config.xml` file.

So, when I add `<include>db.xml</include>`, it should look for `db.xml` in `project/db.xml` and not `htdocs/db.xml`.

Because it checks relative to `htdocs`, I have to use `<include>../project/db.xml</include>` instead, which isn't desirable (in my opinion).

Also, presently, I have to use at least *two* `<include>` tags in order for it to work. So, the following is not enough (will error),

```
<include>../project/db.xml</include>
```

I have to use,

```
<include>../project/db.xml</include>
<include>dummy-value-that-will-not-be-used-at-all</include>
```

**Comment by AnyhowStep** [comment]
My test was just the following in `db.xml`,

```
<config>
    <database>
        <host>localhost</host>
        <username>username</username>
        <password>password</password>
        <database>database</database>
        <quatUser>user</quatUser>
        <quatPassword>pass</quatPassword>
        <name>name</name>
    </database>
</config>
```

**Comment by ridz1208** [comment]
@AnyhowStep 

 - For the first part of your feedback, the intent was for people to use absolute paths rather than relative paths. Regarding the base directory for a relative path... well it's arguable so I'll leave it up to the team, the config class uses itself as the base directory, the loris team tends to assume that the config.xml is under the `project/` directory but this is not the rule (see lines 75 to 87 of `NDB_Config.class.inc`) so the base directory is TBD.

 - For the second part, you are correct I was supposed to change the `toArray()` call to `asArray()` after our last discussion in the meeting, I had forgotten

thx for the feedback

**Comment by driusan** [comment]
I agree that being relative to htdocs isn't very ideal. How much work would be involved in making it relative to the config file which is including it?

**Comment by ridz1208** [comment]
@AnyhowStep 

as discussed with @driusan I foolowed the previously determined "acceptable locations" I reused the code of the config.xml file path location

Note: I also added the possibility to feed the absolute path directly, @johnsaigle not sure if this could be a security issue in any way ? just making sure

**Comment by ridz1208** [comment]
@AnyhowStep please retest !

**Comment by ridz1208** [comment]
@kongtiaowang @driusan 

I give up on trying to understand the simple mind of Travis. I vote we should fire him !!

Go home Travis, you'r drunk !

**Comment by johnsaigle** [comment]
@ridz1208 Should be fine from a security perspective. It's not really possible to do anything nasty with this unless someone has access to `config.xml` on the backend. 👍

**Comment by ridz1208** [comment]
@kongtiaowang did you determine the cause ???

**Comment by ridz1208** [comment]
@kongtiaowang changes made

File: php/libraries/NDB_Config.class.inc [hunk 1]
```diff
@@ -113,9 +108,50 @@ class NDB_Config
...
+            // Standard Unix filesystem layout
+            $file = "/etc/loris/$pathToFile";
+        } else if (file_exists("/usr/local/etc/loris/$pathToFile")) {
+            // Standard Unix filesystem layout, option 2
+            $file = "/usr/local/etc/loris/$pathToFile";
+        } else if (file_exists($pathToFile)) {
+            // Absolute path given
+            $file = $pathToFile;
+        } else {
+            throw new LorisException(
```
Comment (driusan) [inline php/libraries/NDB_Config.class.inc]: wouldn't a ConfigurationException be more appropriate here?

File: php/libraries/Utility.class.inc [hunk 2]
```diff
@@ -558,7 +558,7 @@ class Utility
      */
     static function asArray($var)
     {
-        if (!is_array($var)) {
+        if (!is_array($var) && !empty($var)) {
```
Comment (driusan) [inline php/libraries/Utility.class.inc]: Why is this changed? This seems likely to have unexpected impacts in different parts of the code. Can't you add the empty check in the place that you're using it to be safe?

Comment (ridz1208) [inline php/libraries/Utility.class.inc]: @driusan  this still should be changed though otherwise you return `array([0] => )`

dont you think ?

Comment (driusan) [inline php/libraries/Utility.class.inc]: Another PR can propose changing it, I just don't think it should be done so close to a release because of how far reaching the change is in terms of how low level it is and how many modules and instruments that use the function.
